resources/domain_join.rb

```diff
@@ -120,7 +120,78 @@
   end
 end
 
-# TODO: implement domain leave actions
-# action :leave do
-#
-# end
+action :leave do
+    
+  reboot 'Restart Computer' do
+    action :nothing
+  end
+
+  case node['os']
+  when 'windows'
+    warning_caption = 'Chef is leaving the domain'
+    warning_text = 'The chef cookbook ad-join is currently in the middle of leaving the domain, and the server is about to be restarted'
+    warning_key = 'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System'
+
+    # Display a warning incase anyone remotes into the machine before all reboots have finished
+    # http://www.techrepublic.com/blog/windows-and-office/adding-messages-to-windows-7s-logon-screen/
+    registry_key 'warning' do
+      key warning_key
+      values [
+        { name: 'legalnoticecaption', type: :string, data: warning_caption },
+        { name: 'legalnoticetext', type: :string, data: warning_text }
+      ]
+      only_if { node['ad-join']['windows']['visual_warning'] == true }
+      action :nothing
+    end
+
+    # Installs task for chef-client run after reboot, needed for ohai reload
+    windows_task 'chef ad-join leave' do
+      task_name 'chef ad-join leave' # http://bit.ly/1WDZ1kn
+      user 'SYSTEM'
+      command 'chef-client -L C:\chef\chef-ad-join.log'
+      run_level :highest
+      frequency :onstart
+      only_if { node['kernel']['cs_info']['domain_role'].to_i == 1 || node['kernel']['cs_info']['domain_role'].to_i == 3 }
+      notifies :create, 'registry_key[warning]', :immediately # http://bit.ly/1WDZ1kn
+      action :create
+    end
+
+    # Modify the start time to make sure GP doesn't set task into future
+    # https://github.com/NetDocuments/ad-join-cookbook/issues/13
+    # schedtask.exe won't allow this to be combined with task creation
+    windows_task 'chef ad-join leave' do
+      task_name 'chef ad-join leave' # http://bit.ly/1WDZ1kn
+      start_day '06/09/2016'
+      start_time '01:00'
+      only_if { node['kernel']['cs_info']['domain_role'].to_i == 1 || node['kernel']['cs_info']['domain_role'].to_i == 3 }
+      action :change
+    end
+
+    powershell_script 'ad-join-leave' do
+      code <<-EOH
+      $adminname = "#{domain}\\#{domain_user}"
+      $password = '#{domain_password}' | ConvertTo-SecureString -asPlainText -Force
+      $credential = New-Object System.Management.Automation.PSCredential($adminname,$password)
+
+      Remove-Computer -UnjoinDomainCredential $credential -Force -PassThru
+      EOH
+      only_if { node['kernel']['cs_info']['domain_role'].to_i == 1 || node['kernel']['cs_info']['domain_role'].to_i == 3 }
+      notifies :reboot_now, 'reboot[Restart Computer]', :immediately
+    end
+
+    file 'C:/Windows/chef-ad-join.txt' do
```

Comment:
When a machine is unjoined from a domain, will it ever be rejoined? If the leave action deletes c:/windows/chef-ad-join.txt and then chef runs again, it will most likely rejoin it to the domain.